### ISSUE TITLE
New Java sort Algorithm: "Comparison method violates its general contract"

### ISSUE THREAD

**Issue by fin** [tabulapdf/tabula-java#116]
Hey everyone,

had this issue for example in

SpreadsheetExtractionAlgorithm.java:241
and
Cell.java:34

The reason is that transitivity of compareTo is enforced, which is not supported by methods like Rectangle:compareTo and TexChunk:compareTo - which contain multiple comparisons to decide which dimension to compare on.

Solution: don't know yet, maybe compare across all dimensions at once?

Workaround: pass ` -Djava.util.Arrays.useLegacyMergeSort=true` to java (:

**Comment by jazzido** [comment]
Thanks @fin.

We've run into that a couple times. That's why we include our own implementation of Quicksort in the `technology.tabula.Quicksort` class. Can you try replacing the `Collections.sort` call in `Cell.java` with our Quicksort implementation? Otherwise, if you can provide us with a PDF that triggers the bug, I'll implement a fix.

**Comment by jazzido** [comment]
Actually, all the `Collections.sort` calls should be replaced with [`Utils.sort`](https://github.com/tabulapdf/tabula-java/blob/master/src/main/java/technology/tabula/Utils.java#L121)

**Comment by fin** [comment]
in addition to that pull request, I'm happy to send you the pdf causing the exception, but I'd prefer for it not to be published.

**Comment by giorgiga** [comment]
I was trying to work on this one, but the problem seem to be in the logic rather than in the code and I don't know the codebase at all.

Long story short, the issue is that compareTo implementations must be transitive (it should make sense, but see https://docs.oracle.com/javase/7/docs/api/java/lang/Comparable.html#compareTo(T) if needed).

The current implementation, which can be paraphrased as

> If the two boxes overlap by at least .4, order left-to-right,
> otherwise order top-to-bottom.

is transitive for the "normal" case:
<img width="296" alt="1" src="https://user-images.githubusercontent.com/471835/29842707-ebf0b99e-8d09-11e7-99b3-c4d435f93299.png">
but fails when the boxes are like:
<img width="296" alt="2" src="https://user-images.githubusercontent.com/471835/29842730-f8b8a8ee-8d09-11e7-9a7c-772e5376721d.png">
because A and C do not overlap enough vertically and so C is considered to come after B but before A.

Not 100% I've been able to explain myself... here are two test methods [tests.java.txt](https://github.com/tabulapdf/tabula-java/files/1261660/tests.java.txt) (add them into`TestRectangle`).

I'd very gladly fix this thing myself, but I'm not sure what the ordering of rectangles should be... maybe different criteria will be needed in different situations?

For reference, the fact that `Rectangle` implements `Comparable` is used in:

- `Cell.getText(boolean useLineReturns)`
- `RectangleSpatialIndex.contains(Rectangle r)`
- `RectangleSpatialIndex.intersects(Rectangle r)`
- `SpreadsheetDetectionAlgorithm.detect(Page page)`
- `SpreadsheetExtractionAlgorithm.findSpreadsheetsFromCells(List<? extends Rectangle> cells)`
- `SpreadsheetExtractionAlgorithm.extract(Page page, List<Ruling> rulings)`

**Comment by jazzido** [comment]
Hi @giorgiga,

Thanks for taking a look at this. Transitiveness while sorting 2D elements lexicographically is, probably, impossible to achieve. That's why PDFBox and Tabula include the notion of 'tolerance' when comparing elements.

So far, we haven't run in to breaking cases with the code that we have now. In any case, there are long discussions about a related problem in PDFBox's issue tracker: [PDFBOX-1512](https://issues.apache.org/jira/browse/PDFBOX-1512) and [PDFBOX-2966](https://issues.apache.org/jira/browse/PDFBOX-2996)

**Comment by giorgiga** [comment]
I'm not 100% the PDFBOX guys got the issue straight :)

Transitivity of comparison operators is not a matter of complying to Comparator/Comparable contracts or some other java technicality... comparison operators _must_ be transitive because otherwise there is no ordering (in the math sense) to begin with.

Here's another test case that (I hope) will shed light on the issue

``` java
@Test
public void testWellDefinedComparison1() {
	Rectangle a = new Rectangle(2,0,2,2);
	Rectangle b = new Rectangle(1,1,2,2);
	Rectangle c = new Rectangle(0,2,2,2);
	List<Rectangle> l1 = new ArrayList<>(Arrays.asList(b, a, c));
	List<Rectangle> l2 = new ArrayList<>(Arrays.asList(c, b, a));
	QuickSort.sort(l1);
	QuickSort.sort(l2);
	// At this point l1 and l2 should be the same...
	// I mean, that's kinda the whole purpose of sorting, right?
	// Let's check!
	assertEquals(l1.get(0), l2.get(0));
	assertEquals(l1.get(1), l2.get(1));
	assertEquals(l1.get(2), l2.get(2));
}
```

This is what I get
```
java.lang.AssertionError: expected:<technology.tabula.Rectangle[x=2.0,y=0.0,w=2.0,h=2.0,bottom=2.000000,right=4.000000]> 
                           but was:<technology.tabula.Rectangle[x=0.0,y=2.0,w=2.0,h=2.0,bottom=4.000000,right=2.000000]>
```

PS:
Yes, I had to fiddle with the initial sorting of the two lists – that's why they are not a,b,c/c,b,a – to find a configuration that didn't work.
Please not that there is no issue in QuickSort (AFAIK): the existence of a well define order is a prerequisite for any sorting algorithm.

**Comment by jazzido** [comment]
> Transitivity of comparison operators is not a matter of complying to Comparator/Comparable contracts or some other java technicality... comparison operators must be transitive because otherwise there is no ordering (in the math sense) to begin with.

Yes, but as I understand the issue, there are multiple possible "orderings" in a 2D plane, so transitivity is difficult to define properly, right?

The point of using an ordering of 2d rectangles is sorting glyphs in lexicographical order. See PDFBox's [`PDFTextStripper`](https://github.com/apache/pdfbox/blob/bc8deb87969b2a9f5aedadc7dcaa54322fab97cb/pdfbox/src/main/java/org/apache/pdfbox/text/PDFTextStripper.java#L491-L501) and [`TextPositionComparator`](https://github.com/apache/pdfbox/blob/bc8deb87969b2a9f5aedadc7dcaa54322fab97cb/pdfbox/src/main/java/org/apache/pdfbox/text/TextPositionComparator.java)

**Comment by giorgiga** [comment]
Well, it isn't really a problem that there are many possible ordering of rectangles in a 2D plane... after all, integers can be ordered in many ways too: low-to-high, high-to-low or (eg) comparing according to values modulo 10 (to get equivalence classes like  {...,0,10,20,..} < {...,1,11,21,...} < {...,2,12,22,...} etc).

The issue is that the specific comparison operator implemented in `Rectangle.compareTo()` does not define a total ordering of rectangles. It's not like it's impossible to define a total order in 2D (eg: "top-to-bottom then left-to-right" is perfectly ok); the issue resides in the idea that a well-defined order can be preserved when different quantities are considered equal if they fall within a certain "tolerance" (because a tolerance does not create an equivalence class).

Consider floats (in 1D):

``` java
public class Tolerant implements Comparator<Float> {

    private static final float TOLERANCE = .5f;

    @Override public int compare(Float f1, Float f2) {
        return Math.abs(f1 - f2) <= TOLERANCE
             ? 0
             : f1.compareTo(f2);
    }
}
```
Is `{ 1, 1.2, 1.4, 1.6, 1.8, 2 }` already sorted according to the Tolerant comparator?
And is `{ 1, 1.4, 1.6, 2, 1.2, 1.8 }` already sorted?

What is the expected result of sorting `{ 2, 1, 1.4, 1.6, 1.2, 1.8, }`?

I wouldn't know what to replace the current (ill-defined) order with... I guess it will depend from case to case?
Maybe in some cases the call to `sort()` can be replaced with things that aren't a sorting algorithm... one would need to know the intentions behind the various calls to `sort()` in order to say.

**Comment by jazzido** [comment]
Hi @giorgiga,

> Maybe in some cases the call to sort() can be replaced with things that aren't a sorting algorithm... one would need to know the intentions behind the various calls to sort() in order to say.

Yes, that sounds about right. To avoid misinterpretations and misuses of sorting algorithms, we should use something different than the `Comparable` interface.

BTW, I've been looking at your fork, and I really like the refactors that you've made. Are you interested in sending us a pull-request?

**Comment by giorgiga** [comment]
Pull request: #187 (appveyor fails, but so does it in master)

**Comment by jazzido** [comment]
Thanks @giorgiga,

Will review and merge within the next few days.
